src/loinclib/graph.py

```diff
@@ -15,554 +16,617 @@
-        data[SchemaEnum.TYPE_KEY] = self.type_
-        return current_key
-
-
-class PropertyType(SchemaType):
-    def __init__(
-        self,
-        *,
-        type_: StrEnum,
-        property_owner_type: PropertyOwnerType,
-        name: t.Optional[str],
-        description: t.Optional[str],
-        strict: bool,
-        dynamic: bool,
-    ):
-        super().__init__(
-            type_=type_,
-            name=name,
-            description=description,
-            schema=property_owner_type.schema,
-            strict=strict,
-            dynamic=dynamic,
-        )
-        self.property_owner_type = property_owner_type
 
-    def get_value(self, element: Node | Edge) -> t.Any:
-        properties = element.get_properties()
-        return properties.get(self.type_)
+class EdgeType(Type):
+  pass
+
 
-    def set_value(self, element: Node | Edge, value: t.Any) -> None:
-        if value is None:
-            del element.get_properties()[self.type_]
-        else:
-            element.get_properties()[self.type_] = value
+# @dataclass(kw_only=True)
+# class PropertyTypeArgs(TypeArgs):
+#     pass
+
+@dataclass(kw_only=True)
+class PropertyTypeArgs(TypeArgs):
+  pass
+
+
+class PropertyType(Type):
+  pass
+
+
+# todo: should be abstract
+class Handler:
+  def __init__(
+      self,
+      *,
+      # type_: Type,
+      schema: Schema,
+      name: t.Optional[str],
+      description: t.Optional[str],
+      strict: bool,
+      dynamic: bool,
+  ):
+    # self.type_ = type_
+    self.schema = schema
+    self.name = name
+    self.description = description
+    self.strict = strict
+    self.dynamic = dynamic
+
+
+# todo: should be abstract
+class PropertyOwnerHandler(Handler):
+  def __init__(
+      self,
+      *,
+      # type_: Type,
+      schema: Schema,
+      name: t.Optional[str],
+      description: t.Optional[str],
+      strict: bool,
+      dynamic: bool,
+  ):
+    super().__init__(
+        # type_=type_,
+        name=name,
+        description=description,
+        schema=schema,
+        strict=strict,
+        dynamic=dynamic,
+    )
+    self.property_handlers: t.Dict[PropertyType, PropertyHandler] = {}
+
+  def add_property_type(self, property_handler: PropertyHandler):
+    if property_handler.type_ in self.property_handlers:
+      raise ValueError(f"Duplicate property handler: {property_handler}")
+    self.property_handlers[property_handler.type_] = property_handler
+
+  def get_property_handler(self, type_: PropertyType) -> t.Optional[PropertyHandler]:
+
+    prop = self.property_handlers.get(type_, None)
+    if prop:
+      return prop
+    if self.strict:
+      raise ValueError(f"Unknown property type: {type_}")
+    prop = self.schema.create_property_handler(
+        type_=type_, property_owner_handler=self, dynamic=True, strict=self.strict
+    )
+    self.property_handlers[type_] = prop
+    return prop
+
+
+class NodeHandler(PropertyOwnerHandler):
```

Comment:
Just gonna trust you on most of these new additions, rather than in depth review.

Just wondering, how does it do things like `.get_node_by_code()`? I don't see any graph, node, or edge properties in this class. I guess maybe it uses schema? Not sure.